File: PYTHON/Classes/4. my_hash.py
def my_hash(string: str) -> int:
    hash = 0
    for ch in string:
        hash += ord(ch)
    return hash


def test_my_hash():
    assert my_hash('') == 0
    assert my_hash('a') == 97
    assert my_hash('ab') == 97 + 98
    assert my_hash('c') == 99
    assert my_hash('./?C') == my_hash('ß') # collision
    print("All test cases passed!")


test_my_hash()

File: JAVASCRIPT/Classes/4. myHash.js
// Mixing Operations: You might perform additional bitwise operations
// (like XOR or bit shifting) between the characters to achieve
// a better mixing of bits, ensuring a different hash code for similar strings.

const myHash = (string) => {
  let hash = 0;
  for (let i = 0; i < string.length; i++) {
    hash = (hash << 5) ^ (hash >> 27) ^ string.charCodeAt(i);
  }
  return hash & 0x7fffffff;
};

module.exports = myHash;

const test_myHash = () => {
  const hash = myHash('hello');
  console.log(`Hash of 'hello': ${hash}`);

  const hash1 = myHash('olleh');
  console.log(`Hash of 'olleh': ${hash1}`);

  console.log(
    `Collision check: ${hash === hash1 ? 'Collision!' : 'No collision'}`
  );
};

test_myHash();
